octoprint_mrbeam/software_update_information.py

```diff
@@ -19,72 +24,197 @@
     SW_UPDATE_TIER_DEV: "develop",
 }
 
-# add to the display name to modules that should be shown at the top of the list
-SORT_UP_PREFIX = " "
-
+SW_UPDATE_FILE = "update_info.json"
+SW_UPDATE_CLOUD_PATH = "https://mr-beam.org/beamos/config/sw-update-conf"
 
 _logger = mrb_logger("octoprint.plugins.mrbeam.software_update_information")
 
+# sw_update_config = dict()
+
 # Commented constants are kept in case we update more packages from the virtualenv
 # GLOBAL_PY_BIN = "/usr/bin/python2.7"
 # VENV_PY_BIN = sys.executable
 GLOBAL_PIP_BIN = "/usr/local/bin/pip"
 GLOBAL_PIP_COMMAND = (
     "sudo {}".format(GLOBAL_PIP_BIN) if os.path.isfile(GLOBAL_PIP_BIN) else None
 )
-# GLOBAL_PIP_COMMAND = "sudo {} -m pip".format(GLOBAL_PY_BIN) if os.path.isfile(GLOBAL_PY_BIN) else None #  --disable-pip-version-check
-# VENV_PIP_COMMAND = ("%s -m pip --disable-pip-version-check" % VENV_PY_BIN).split(' ') if os.path.isfile(VENV_PY_BIN) else None
+
 BEAMOS_LEGACY_DATE = date(2018, 1, 12)
 
+_softwareupdate_handler = None
+
+class MrBeamSoftwareupdateHandler:
+    CLOUD_FILE = 1
+    LOCAL_FILE = 2
+    REPO_FILE = 3
+    def __init__(self, plugin):
+        self._plugin = plugin
+
+    # @get_thread(daemon=False)
+    @logExceptions
+    def load_update_file_from_cloud(self, localfilemissing=False):
```

Comment:
What is the benefit of the `localfilemissing` variable ? Because the function checks for the file anyway, so it might as well take care of it itself.

I also think this function is trying to do too many things at the same time: downloading a file and managing local copies.

I would recommend to have different functions that
- tries to download the file and can raise an exception
- verify the presence and content of a local file path
- a function that uses the 2 previous ones to dictate the usage flow : if download fails, use the local copy etc...

This will also help you for the testing so variables like `localfilemissing` are more natural and you have more flexibility.

Reply:
This will be used to force to ignore the local file and use the server file, ok I seperate them